drivers/gles2/rasterizer_canvas_gles2.cpp

```diff
@@ -1442,787 +1571,1118 @@ void RasterizerCanvasGLES2::canvas_render_items(Item *p_item_list, int p_z, cons
-					if (storage->frame.current_rt && storage->frame.current_rt->flags[RasterizerStorage::RENDER_TARGET_TRANSPARENT]) {
-						glBlendFuncSeparate(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA, GL_ONE, GL_ONE_MINUS_SRC_ALPHA);
-					} else {
-						glBlendFuncSeparate(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA, GL_ZERO, GL_ONE);
-					}
+		//				glActiveTexture(GL_TEXTURE0 + storage->config.max_texture_image_units - 4);
+		//				RasterizerStorageGLES2::Texture *t = storage->texture_owner.getornull(light->texture);
+		//				if (!t) {
+		//					glBindTexture(GL_TEXTURE_2D, storage->resources.white_tex);
+		//				} else {
+		//					t = t->get_ptr();
 
-				} break;
-				case RasterizerStorageGLES2::Shader::CanvasItem::BLEND_MODE_ADD: {
+		//					glBindTexture(t->target, t->tex_id);
+		//				}
 
-					glBlendEquation(GL_FUNC_ADD);
-					if (storage->frame.current_rt && storage->frame.current_rt->flags[RasterizerStorage::RENDER_TARGET_TRANSPARENT]) {
-						glBlendFuncSeparate(GL_SRC_ALPHA, GL_ONE, GL_SRC_ALPHA, GL_ONE);
-					} else {
-						glBlendFuncSeparate(GL_SRC_ALPHA, GL_ONE, GL_ZERO, GL_ONE);
-					}
+		//				glActiveTexture(GL_TEXTURE0);
+		//				_canvas_item_render_commands(ci, NULL, reclip, material_ptr); //redraw using light
 
-				} break;
-				case RasterizerStorageGLES2::Shader::CanvasItem::BLEND_MODE_SUB: {
+		//				state.using_light = NULL;
+		//			}
 
-					glBlendEquation(GL_FUNC_REVERSE_SUBTRACT);
-					if (storage->frame.current_rt && storage->frame.current_rt->flags[RasterizerStorage::RENDER_TARGET_TRANSPARENT]) {
-						glBlendFuncSeparate(GL_SRC_ALPHA, GL_ONE, GL_SRC_ALPHA, GL_ONE);
-					} else {
-						glBlendFuncSeparate(GL_SRC_ALPHA, GL_ONE, GL_ZERO, GL_ONE);
-					}
-				} break;
-				case RasterizerStorageGLES2::Shader::CanvasItem::BLEND_MODE_MUL: {
-					glBlendEquation(GL_FUNC_ADD);
-					if (storage->frame.current_rt && storage->frame.current_rt->flags[RasterizerStorage::RENDER_TARGET_TRANSPARENT]) {
-						glBlendFuncSeparate(GL_DST_COLOR, GL_ZERO, GL_DST_ALPHA, GL_ZERO);
-					} else {
-						glBlendFuncSeparate(GL_DST_COLOR, GL_ZERO, GL_ZERO, GL_ONE);
-					}
+		//			light = light->next_ptr;
+		//		}
+
+		//		if (light_used) {
+
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::USE_LIGHTING, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::USE_SHADOWS, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_NEAREST, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF3, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF5, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF7, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF9, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF13, false);
+
+		//			state.canvas_shader.bind();
+
+		//			ris.last_blend_mode = -1;
+		//		}
```

Comment:
I think just delete the commented out sections, From what I can tell, you have already pulled out the necessary pieces

Reply:
Yeah I was expecting we'd decide to do that, we can always look at the old version for reference.

In terms of maintenance, this is one of the most error prone areas actually. The logic in try_join must exactly match the logic in the renderer, otherwise you will get subtle rendering bugs as a result of joining items that shouldn't have been joined. In one of the other versions of the PR, I tried putting this all in the same uber function. However it does end up quite messy.

Reply:
I think this is fine for now. I have a feeling that if we add proper state tracking we can simplify this and the main render function enough that the duplication won't be so bad.